src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+    }
+
+    /** Read structures for the template bases, one read structure per read (ex. r1, r2, i7, and i5). */
+    private val templateReadStructures = readStructures.map { readStructure =>
+      readStructure.templateStructure(resetOffsets=false)
+    }
+
+    /** Read structures for the sample barcodes, one read structure per read (ex. r1, r2, i7, and i5). */
+    private val sampleBarcodeReadStructures = readStructures.map { readStructure =>
+     readStructure.sampleBarcodeStructure(resetOffsets=false)
+    }
+
+    /** Read structures for the molecular barcodes, one read structure per read (ex. r1, r2, i7, and i5). */
+    private val molecularBarcodeReadStructures = readStructures.map { readStructure =>
+      readStructure.molecularBarcodeStructure(resetOffsets=false)
+    }
+
+    /** Sample barcode to its associate metric */
+    private val barcodeToMetrics: Map[String, SampleBarcodeMetric] = {
+      val metricMap = new mutable.LinkedHashMap[String, SampleBarcodeMetric]()
+      sampleSheet.foreach { sample =>
+        val barcode: String = sample.sampleBarcode match {
+          case Some(b) => b.concatenatedBarcode
+          case None    => throw new IllegalArgumentException(s"Sample with id '${sample.sampleId}' did not have a sampleBarcode")
+        }
+        val libraryId: String = sample.libraryId match {
+          case Some(id) => id
+          case None     => throw new IllegalArgumentException(s"Sample with id '${sample.sampleId}' did not have a library id")
+        }
+        metricMap.put(barcode, SampleBarcodeMetric(barcodeName=sample.sampleName, libraryName=libraryId, barcode=barcode))
+      }
+      metricMap.put(noMatchBarcode, SampleBarcodeMetric(UnmatchedSampleId, UnmatchedSampleId, noMatchBarcode))
+      metricMap.toMap
+    }
+
+    def hasNext: Boolean = input.hasNext
+
+    def next(): AssignedRecords = {
+      if (!hasNext) throw new NoSuchElementException("Calling next() when hasNext is false")
+
+      val fastqRecords                        = input.next()
+      val sampleBarcodeReadBases: Seq[String] = getSampleBarcode(fastqRecords).filter(_.nonEmpty)
+      val sampleOrdinal: Int                  = getSampleOrdinalFromSampleBarcode(sampleBarcodeReadBases.mkString, sampleSheet).getOrElse(headers.length)
+      val header: SAMFileHeader               = headers(sampleOrdinal-1)
+      val sampleId: String                    = header.getReadGroups.get(0).getId
+      val molecularBarcodes: Seq[String]      = getMolecularBarcodes(fastqRecords).filter(_.nonEmpty)
+      val pairedEnd: Boolean                  = readStructures.count(_.template.nonEmpty) > 1
+
+      val records = fastqRecords.zipWithIndex.flatMap { case (record, recordIndex) =>
+        val readStructure = templateReadStructures(recordIndex)
+        if (readStructure.isEmpty) {
+          None
+        }
+        else {
+          if (1 < recordIndex) throw new IllegalArgumentException("Found more than two ends with template bases")
```

Comment:
This is a little tenuous, since it's only testing if you found template bases in recordIndex 2 or higher - but maybe it's possible (though irregular) that one of the first two reads didn't contain template bases but a later read did?